### PR TITLE
[NETBEANS-4826] Fixing java indexing in Groovy

### PR DESCRIPTION
**Pull request by ppisl**
This will turn back on indexing of Groovy files in Java. This means that Groovy classes can now be offered in java source files.  The Groovy file is used to create virtual Java source code that is indexed and stored in the java index. This will not only enable indexing, but features like code completion, hyperlinks, goto navigation should also work. 

More fixes were needed to get this working again. There are also a few fixes that improve the current staff, such as:

- static fields and methods are offered as fields and methods in import statements code completion - they were previously offered as classes in code completion. 
- correcting types of constructors that are strored in Groovy Index
- JavaElementHandles now are created with the fully qualifed names, there were created only with simple names. 
- adding @Trusted annotation for ForwardingJavaFileObject implementations

### CODE REVIEW

File: groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/api/completion/util/CompletionContext.java [hunk 1]
```diff
@@ -298,19 +298,28 @@ private CaretLocation getCaretLocationFromRequest() {
...
                 classDefBeforePosition = true;
                 break;
             } else if (t.id() == GroovyTokenId.LITERAL_package) {
                 afterPackagePosition = true;
                 break;
+            } else if (canBeImport && t.id() == GroovyTokenId.LITERAL_import) {
+                return CaretLocation.INSIDE_IMPORT;
+            }
+            
+            if (canBeImport && !(t.id() == GroovyTokenId.DOT || t.id() == GroovyTokenId.IDENTIFIER
```
Comment (sdedic) [inline groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/api/completion/util/CompletionContext.java:320]: Q: how will this react in context like `Whatever.class.` ?

File: groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/completion/TypesCompletion.java [hunk 2]
```diff
@@ -313,12 +314,24 @@ private void addToProposalUsingFilter(Set<TypeHolder> alreadyPresent, TypeHolder
...
             return;
         }
+        
+        if (type.getHandle() != null
+                &&!(type.getHandle().getKind().isClass() || type.getHandle().getKind().isInterface())
+                && request.location != CaretLocation.INSIDE_IMPORT) {
+            return;
+        }
 
+        String ownerFQN = GroovyUtils.stripClassName(fqnTypeName);
```
Comment (sdedic) [inline groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/completion/TypesCompletion.java:324]: Use `GroovyUtils.getPackageName()`, returns name w/o trailing dot.

File: groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/completion/TypesCompletion.java [hunk 3]
```diff
@@ -313,12 +314,24 @@ private void addToProposalUsingFilter(Set<TypeHolder> alreadyPresent, TypeHolder
...
+        if (!ownerFQN.isEmpty() && ownerFQN.endsWith("."))  {
+            ownerFQN = ownerFQN.substring(0, ownerFQN.length() - 1);
+        }
+        
         // We are dealing with prefix for some class type
         JavaElementHandle jh = null;
         if (type.getHandle() != null) {
-            jh = new JavaElementHandle(fqnTypeName, typeName, type.getHandle(), Collections.emptyList(), Collections.emptySet());
+            jh = new JavaElementHandle(typeName, ownerFQN, type.getHandle(), Collections.emptyList(), Collections.emptySet());
```
Comment (sdedic) [inline groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/completion/TypesCompletion.java:332]: Uh-oh ... this was a stupid bug; thanks.

**Review by sdedic — COMMENTED** [review]
Please leave a comment in the TypeHolder, that it is actually being (mis)used also for static fields and methods, despite its name.

**Review by dbalek — APPROVED** [review]

File: groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/api/completion/util/CompletionContext.java (hunk 1 above)
Comment (ppisl) [reply groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/api/completion/util/CompletionContext.java:320]: Can we have import statement like `import Whatever.class` ?

File: groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/completion/TypesCompletion.java (hunk 2 above)
Comment (ppisl) [reply groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/completion/TypesCompletion.java]: Done

**Review by sdedic — APPROVED** [review]

**Pull request merged**